server/tpm2.go

```diff
@@ -4,10 +4,125 @@
 package main
 
 import (
+	"io"
+
 	"github.com/google/go-tpm/tpm2"
 	"github.com/google/go-tpm/tpmutil"
+	"github.com/sirupsen/logrus"
 )
 
+// TCG TPM v2.0 Provisioning Guidance, section 7.8:
+// https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf
+const SRK_HANDLE tpmutil.Handle = 0x81000001
+
+// TCG TPM v2.0 Provisioning Guidance, section 7.5.1:
+// https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf
+var SRK_TEMPLATE = tpm2.Public {
+	Type: tpm2.AlgRSA,
+	NameAlg: tpm2.AlgSHA256,
+	Attributes: tpm2.FlagFixedTPM | tpm2.FlagFixedParent | tpm2.FlagSensitiveDataOrigin | tpm2.FlagUserWithAuth | tpm2.FlagRestricted | tpm2.FlagDecrypt | tpm2.FlagNoDA,
+	AuthPolicy: nil,
+	RSAParameters: &tpm2.RSAParams {
+		Symmetric: &tpm2.SymScheme {
+			Alg: tpm2.AlgAES,
+			KeyBits: 128,
+			Mode: tpm2.AlgCFB,
+		},
+		KeyBits: 2048,
+		ModulusRaw: make([]byte, 256),
+	},
+}
+
+/*
+	Returns a handle to the TPM Storage Rook Key (SRK) if it's already present
+	in it's non volatile memory.
+	Else, a new SRK is created using the default template, and is persisted
+	in the TPM Non Volatile memory, at it's default location. A handle to the newly
+	created key is then returned
+*/
+func TPM2_LoadSRK(rwc io.ReadWriteCloser) (tpmutil.Handle, error) {
+	handles, _, err := tpm2.GetCapability(rwc, tpm2.CapabilityHandles, 1000, uint32(tpm2.HandleTypePersistent) << 24)
```

Comment:
Where do the count=1000 and <<24 come from?

Reply:
`1000` is an arbitrary value specifying the maximum object number `GetCapability` can return. I agree it deserves at least a const qualified name.

I don't remember where I found the `<< 24` trick, but I now I saw it somewhere (in a call to this same function). It is because depending on the capability given to `GetCapability`, the property argument can differ, but is an `uint32` whatever it means. For the `CapabilityHandle` capability, the tpm2 library provides constants, but they are `uint8`. The compiler enforces that an `uint32` is passed to the function, so I have either to cast or to convert the provided constant.

My guess is that when converting it to an uint32, the `uint8` value is implicitly shifted 24 bytes to the right (because intel machines are little endian), and that they cast the value back to an `uint8` inside the function, which results to a `0` if we don't shift it back.

If I'm right, an option would be to cast by myself the `uint8` to an `uint32`, instead of convert it. But casting isn't supported and I'd need dirty tricks as shown here: https://medium.com/a-journey-with-go/go-cast-vs-conversion-by-example-26e0ef3003f0

Reply:
Let's add a comment to explain the situation:
// Shift the handle type into the MSO according to the specification: "TPM_CAP_HANDLES – Returns a list of all of the handles within the handle range of the property parameter. The range of the returned handles is determined by the handle type (the most-significant octet (MSO) of the property)."
// https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part3_Commands_pub.pdf Section 30.2 TPM2_GetCapability